### PR TITLE
Stabilize THIR unsafeck

### PR DESCRIPTION
**Pull request by matthewjasper**
- Make `-Zthir-unsafeck`=on, the default
- Update some compiler comments/names to treat THIR unsafeck as the default.
- Union patterns are now unsafe unless the field is matched to a wildcard pattern.
- Fixes an issue with safety checking inline constant patterns.

### CODE REVIEW

**Comment by lqd** [comment]
A perf run would be nice as well: the compiler perf WG had plans to help with this switch because THIR unsafeck was a bit faster than MIR unsafeck.

**Comment by matthewjasper** [comment]
@craterbot check

**Comment by lqd** [comment]
@matthewjasper I also wonder if we should stabilize and enable THIR unsafeck first, and only remove MIR unsafeck after a period of time has passed, to allow for feedback and more testing? That way we'd still be able to easily flip a flag back to MIR unsafeck for beta/stable branches if we needed to.

I guess that can be discussed during FCP.

**Comment by Mark-Simulacrum** [comment]
@craterbot retry-report

**Comment by matthewjasper** [comment]
The wlcs regression is essentially this:

```rust
#![deny(unsafe_op_in_unsafe_fn)]

unsafe fn deref(x: *mut i32) {
    let y = std::ptr::addr_of!(*x);
}
```

I don't think there's much that we can do here. The lint should fire here, and only doesn't in the current unsafeck because of MIR only having a Span for `&raw mut *x`, which doesn't get linted because it's in a macro expansion. At least this is a lint so will not prevent compiling as a dependency.

I'll check whether any of the other regressions are legitimate later.

**Comment by matthewjasper** [comment]
@craterbot run mode=check-only crates=https://crater-reports.s3.amazonaws.com/pr-117673/retry-regressed-list.txt

**Comment by lqd** [comment]
This is going to be a short run:
@craterbot p=1

**Comment by matthewjasper** [comment]
Nominating for T-lang discussion. This is mostly an internal change, but there are a couple of differences between THIR and MIR unsafeck that are visible to users.

- Union patterns are considered to access fields (and so need `unsafe`) in more cases, crater could not find any code this breaks
```rust
let U { f } = u;     // access for both checkers
let U { f: () } = u; // only an access in THIR unsafeck
let U { f: _ } = u;  // not an access in either cheker
```
- Some error spans are slightly better in THIR unsafeck
- Some _lint_ spans are slightly different in THIR unsafeck. Because `unsafe_op_in_unsafe_fn` is not currently a backwards compatibility lint, nor reported in external macros, this can change whether it's reported. This can break code with `deny(unsafe_op_in_unsafe_fn)` when not compiled with `--cap-lints`. Crater found one place where this happened.

**Comment by traviscross** [comment]
@rustbot labels +T-opsem

This was discussed in the T-lang triage meeting today.  There was interest in getting feedback from T-opsem about this.  @nikomatsakis may leave a more extensive comment.

cc @rust-lang/opsem

**Comment by RalfJung** [comment]
Great to see this project nearing completion. :)

IIUC, this makes strictly *more* code unsafe than was unsafe before? If so, no concerns from my side. ;)  (But I agree with the concerns raised by @scottmcm above.)

I assume all the tests pass so the changes to unsafeck that happened since the THIS unsafeck project started are all incorporated? For instance
- https://github.com/rust-lang/rust/pull/103208, https://github.com/rust-lang/rust/pull/103208 (these change MIR building but that has effects on unsafeck as well)
- https://github.com/rust-lang/rust/pull/100081
- https://github.com/rust-lang/rust/pull/97995

**Comment by saethlin** [comment]
I searched the crater run for the error message that would appear if the newly-failing scenario were here. No results.

So no concerns from me.

**Comment by Nilstrieb** [comment]
Can we keep MIR unsafeck in this PR and only flip the switch and then remove MIR unsafeck only once it hits beta/stable? That should make it easier to revert in case there are bad regressions that can't be fixed easily, which I don't expect, but could always happen.

**Comment by pnkfelix** [comment]
We discussed this in the T-Lang triage meeting today.

The team members present were are all on board for stabilizing THIR unsafeck, including the semantic changes described by @matthewjasper in their [comment above](https://github.com/rust-lang/rust/pull/117673#issuecomment-1812659390).

We are assuming that this PR will be updated in the manner described such that it focuses solely on stabilizing THIR unsafeck, and will not include the changes to MIR construction. (We also expect that the removal of MIR unsafeck will be delayed for a later PR, e.g. so that we can trivially flip back to MIR unsafeck during the beta evaluation in case some problem is found with THIR unsafeck during that time.)

@rfcbot fcp merge

**Comment by pnkfelix** [comment]
omg sorry about that.

@rfcbot fcp cancel

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

As discussed above, the consensus was to do this subject to the promised changes.  This is now in FCP.  Let's remove the nomination.

File: compiler/rustc_mir_build/src/check_unsafety.rs [hunk 1]
```diff
@@ -148,6 +148,8 @@ impl<'tcx> UnsafetyVisitor<'_, 'tcx> {
                 hir_context,
                 safety_context,
                 warnings: self.warnings,
+                in_union_destructure: false,
+                inside_adt: false,
                 ..*self
```
Comment (cjgillot) [inline compiler/rustc_mir_build/src/check_unsafety.rs:153]: Should we make this fully explicit?

**Comment by matthiaskrgr** [comment]
The pr body says `Removes -Zthir-unsafeck` but it seems that
`    -Z                                 thir-unsafeck=val -- use the THIR unsafety checker (default: yes)` is still a thing, is this intended?

**Comment by nnethercote** [comment]
Perf improvements greatly outweigh the tiny number of regressions.

@rustbot label: +perf-regression-triaged

**Pull request merged**